### PR TITLE
helpers.recursive_update: merging lists of Dictionaries

### PR DESCRIPTION
**Pull request by ngr**
helpers.recursive_update: Now supports merging lists of Dictionaries preserving unique

### CODE REVIEW

**Comment by ngr** [comment]
Hi guys. Please explore the slightly modified behaviour of `sosw.helpers.recursive_update`.
In case some element is a list of dictionaries it will filter our completely unique ones.
The rest of the behaviour should remain the same.

File: sosw/components/helpers.py [hunk 1]
```diff
@@ -798,19 +797,29 @@ def recursive_update(d: Dict, u: Mapping) -> Dict:
...
+                    new[k] = None
+
+                # If types are not hashable we still try to deal with them as if Dictionaries.
+                # In this case we filter unique ones by JSON values and them unfold them back and reconstruct Dicts.
+                if not new[k]:
+                    try:
+                        jsons = set(json.dumps(sorted(x.items())) for x in nv)
+                        new[k] = [dict(json.loads(x)) for x in jsons]
+                    except (TypeError, AttributeError):
+                        # In not all values of iterable are hashable and not Dictionaries we just merge them as is.
```
Comment (DrugBug) [inline sosw/components/helpers.py:821]: 'In' --> 'If'

File: sosw/components/helpers.py [hunk 2]
```diff
@@ -798,19 +797,29 @@ def recursive_update(d: Dict, u: Mapping) -> Dict:
...
-                    # In this case we just merge lists as is.
-                    new[k] = nv
+                    new[k] = None
+
+                # If types are not hashable we still try to deal with them as if Dictionaries.
+                # In this case we filter unique ones by JSON values and them unfold them back and reconstruct Dicts.
+                if not new[k]:
+                    try:
+                        jsons = set(json.dumps(sorted(x.items())) for x in nv)
+                        new[k] = [dict(json.loads(x)) for x in jsons]
```
Comment (nimirium) [inline sosw/components/helpers.py:819]: umm but what if the order matters?

**Comment by nimirium** [comment]
It's not bad, but it's likely that order of the dictionaries in the list matters!

**Comment by ngr** [comment]
@nimirium, I have just reread the code. The behaviour of the ORDERING doesn't change from the previous behaviour of merging lists, tuples and sets. As long as the function supports mixing these three types the order was never guaranteed.

As for `sorted()` function that probably confused you - this is used only for reordering KEYS in dictionaries. Preserving order of keys in a dict is a recent "side-effect", not a "feature" of Python. And it definitely should not be guaranteed during dictionary transformations.

**Comment by nimirium** [comment]
@ngr 
```
jsons = set(json.dumps(sorted(x.items())) for x in nv)
```
set() mixes up the order of the dictionaries in the list. 
The order of the items in the list might be important.

Although that's what you do to the other data types, so I guess it's consistent...

**Comment by ngr** [comment]
> Although that's what you do to the other data types, so I guess it's consistent...

Indeed.
> The behaviour of the ORDERING doesn't change from the previous behaviour of merging lists,
> tuples and sets. As long as the function supports mixing these three types the order was never
> guaranteed.

**Pull request merged**